src/wasm/lang.rs

```diff
@@ -0,0 +1,47 @@
+use wasm_bindgen::prelude::*;
+
+/// Flux provides an API for transforming, formatting, and checking syntax of flux source code.
+#[wasm_bindgen]
+pub struct Flux {
+    ast: flux::ast::Package,
+}
+
+#[wasm_bindgen]
+impl Flux {
+    /// Create a new Flux object from a raw flux string.
+    #[wasm_bindgen(constructor)]
+    pub fn new(script: &str) -> Self {
+        let mut parser = flux::parser::Parser::new(script);
+        let parsed = parser.parse_file("".into());
+
+        Self { ast: parsed.into() }
+    }
+
+    /// Format the flux.
+    ///
+    /// In the event that the flux is invalid syntax, an Err will be returned,
+    /// which will translate into a JavaScript exception being thrown.
+    #[wasm_bindgen]
+    pub fn format(&self) -> Result<String, String> {
+        // XXX: rockstar (1 Apr 2022) - This currently only supports a single file package. It should
+        // take a parameter for the file to format.
+        flux::formatter::convert_to_string(&self.ast.files[0])
+            .map_err(|err| format!("{}", err))
+    }
+
+    /// Check that the flux is valid. This function does a semantic check, which will check types and builtin
+    /// function signatures, which can't be checked via a base AST check.
+    #[wasm_bindgen]
+    pub fn check(&self) -> bool {
```

Comment:
I thought about this a bunch, and I think it's mostly because I wanted to create the most minimal api possible so that we could determine _how_ we wanted the errors to show up, e.g. we may not actually need the errors, just whether or not it's valid.

While responding to this, I think my response here is to just rename this function to `is_valid`, and a future `check` function could return a list of errors.

Reply:
can we have both? with the `is_valid` reference the `check` and match response to bool?

Reply:
What format should the errors return in `check`? How will the error be consumed? I'm hesitant to expose that until we have a good handle what's needed in the error. We can't just expose semantic check errors as an object, since those aren't a public interface (and shouldn't be). What would be relevant information to convey in an error object?

Reply:
Our goal for the semantic check is to have enough details for the UI developer to debug, as well something general that we can surface to the user. The UI team can do the work to iterate on the second point (what to surface). Product has agreed they can extract something useful from this requested error payload.

For us devs to effectively debug, may we have:
* Positional information
* ErrorKind. (This can be any error enum variant which provides useful information.)
* Source.
* Message, if available in the error.

Example from the flux_check_bad_type:
```
{
	valid: false,
	errors: [{
		location: { line: 1, column: 14 },
		source: “1”,
		type: “CannotUnifyArgument”,
	}]
}
```

Example from the flux_check_invalid_syntax:
```
{
	valid: false,
	error: [{
		location: { line: 1, column: 1 },
		source: "from(bucket: 1 |> range(start: -15m)",
		type: “InvalidAST”,
		message: "expected RPAREN, got EOF”,
 	}, {
		location: { line: 1, column: 14 },
		source: “1”,
		type: “CannotUnifyArgument”,
	} … rest of errors…]
}
```

Those ^^ are only illustrative examples, open to discussion. And we understand that the error API is still undergoing development, so we would need to make sure to have a conversation if/when we decide upon any specific kind-match handling. But for now, it’s okay for some drift in the error API. It’s more important that we get actionable information.

What do you think?

Reply:
Had a very productive convo with @rockstar . Will summarize here (please feel free to also add context/clarification).

The errors I have been viewing within the semantic check(), are based on error logging which is not part of the compressed wasm binary. Therefore a simple surfacing of these errors is not possible, without having to explode the binary size.

Instead, (and as part of a longterm plan which @rockstar already has), the surfacing of errors will require more work and public API design to selectively provide this information -- and in a more size efficient fashion.

Therefore, for now, we have agreed to only provide the boolean on semantic check.